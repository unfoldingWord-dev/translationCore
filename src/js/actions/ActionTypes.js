/* istanbul lignore next */
const ActionTypes = {
  CHANGE_ONLINE_STATUS: "CHANGE_ONLINE_STATUS",
  START_LOADING: "START_LOADING",
  SHOW_MODAL_CONTAINER: "SHOW_MODAL_CONTAINER",
  SELECT_MODAL_TAB: "SELECT_MODAL_TAB",
  LOGOUT_USER: "LOGOUT_USER",
  LOGIN_USER: "LOGIN_USER",
  SET_TOOLS_METADATA: "SET_TOOLS_METADATA",
  GET_RECENT_PROJECTS: "GET_RECENT_PROJECTS",
  RECIEVE_REPOS: "RECIEVE_REPOS",
  IMPORT_LINK: "IMPORT_LINK",
  TOGGLE_LOADER_MODAL: "TOGGLE_LOADER_MODAL",
  SELECT_MODAL_SECTION: "SELECT_MODAL_SECTION",
  SHOW_POPOVER: "SHOW_POPOVER",
  CLOSE_POPOVER: "CLOSE_POPOVER",
  SAVE_TOOL_VIEW: "SAVE_TOOL_VIEW",
  // Resources Actions
  ADD_NEW_BIBLE_TO_RESOURCES: "ADD_NEW_BIBLE_TO_RESOURCES",
  ADD_TRANSLATIONHELPS_ARTICLE: "ADD_TRANSLATIONHELPS_ARTICLE",
  ADD_LEXICON_ENTRY: "ADD_LEXICON_ENTRY",
  // Other actions that need organized
  SET_SAVE_PATH_LOCATION: "SET_SAVE_PATH_LOCATION",
  UPDATE_PROGRESS: "UPDATE_PROGRESS",
  TOGGLE_REMINDER: "TOGGLE_REMINDER",
  SET_REMINDER: "SET_REMINDER",
  CHANGE_CURRENT_CONTEXT_ID: "CHANGE_CURRENT_CONTEXT_ID",
  ADD_COMMENT: "ADD_COMMENT",
  ADD_VERSE_EDIT: "ADD_VERSE_EDIT",
  ADD_GROUP_DATA: "ADD_GROUP_DATA",
  CHANGE_SELECTIONS: "CHANGE_SELECTIONS",
  CLEAR_CURRENT_TOOL_DATA: "CLEAR_CURRENT_TOOL_DATA",
  STORE_MANIFEST: "STORE_MANIFEST",
  STORE_SAVE_LOCATION: "STORE_SAVE_LOCATION",
  SET_CURRENT_TOOL_NAME: "SET_CURRENT_TOOL_NAME",
  SET_CURRENT_TOOL_TITLE: "SET_CURRENT_TOOL_TITLE",
  LOAD_GROUPS_INDEX: "LOAD_GROUPS_INDEX",
  LOAD_GROUPS_DATA_FROM_FS: "LOAD_GROUPS_DATA_FROM_FS",
  SET_REMINDERS_IN_GROUPDATA: "SET_REMINDERS_IN_GROUPDATA",
  SET_INVALIDATION_IN_GROUPDATA: "SET_INVALIDATION_IN_GROUPDATA",
  TOGGLE_REMINDERS_IN_GROUPDATA: "TOGGLE_REMINDERS_IN_GROUPDATA",
  TOGGLE_SELECTIONS_IN_GROUPDATA: "TOGGLE_SELECTIONS_IN_GROUPDATA",
  TOGGLE_VERSE_EDITS_IN_GROUPDATA: "TOGGLE_VERSE_EDITS_IN_GROUPDATA",
  TOGGLE_COMMENTS_IN_GROUPDATA: "TOGGLE_COMMENTS_IN_GROUPDATA",
  CLEAR_PREVIOUS_GROUPS_DATA: "CLEAR_PREVIOUS_GROUPS_DATA",
  CLEAR_PREVIOUS_GROUPS_INDEX: "CLEAR_PREVIOUS_GROUPS_INDEX",
  CLEAR_CONTEXT_ID: "CLEAR_CONTEXT_ID",
  UPDATE_TARGET_VERSE: "UPDATE_TARGET_VERSE",
  SHOW_LOADING_CIRCLE: "SHOW_LOADING_CIRCLE",
  HIDE_LOADING_CIRCLE: "HIDE_LOADING_CIRCLE",
  GOGS_SERVER_ERROR: "GOGS_SERVER_ERROR",
  OPEN_ALERT_DIALOG: "OPEN_ALERT_DIALOG",
  OPEN_OPTION_DIALOG: "OPEN_OPTION_DIALOG",
  CLOSE_ALERT_DIALOG: "CLOSE_ALERT_DIALOG",
  LOGIN_LOCAL_USER: "LOGIN_LOCAL_USER",
  RESET_PROJECT_DETAIL: "RESET_PROJECT_DETAIL",
  CLEAR_RESOURCES_REDUCER: "CLEAR_RESOURCES_REDUCER",
  UPDATE_ONLINE_MODE: "UPDATE_ONLINE_MODE",
  // Online import actions
  SET_REPOS_DATA: "SET_REPOS_DATA",
  // project actions
  SET_PROJECT_PROGRESS_FOR_TOOL: "SET_PROJECT_PROGRESS_FOR_TOOL",
  SET_GL_FOR_TOOL: "SET_GL_FOR_TOOL",
  // Home
  TOGGLE_HOME_VIEW: "TOGGLE_HOME_VIEW",
  TOGGLE_WELCOME_SPLASH: "TOGGLE_WELCOME_SPLASH",
  GO_TO_STEP: "GO_TO_STEP",
  OPEN_PROJECTS_FAB: "OPEN_PROJECTS_FAB",
  CLOSE_PROJECTS_FAB: "CLOSE_PROJECTS_FAB",
  GET_MY_PROJECTS: "GET_MY_PROJECTS",
  OPEN_ONLINE_IMPORT_MODAL: "OPEN_ONLINE_IMPORT_MODAL",
  CLOSE_ONLINE_IMPORT_MODAL: "CLOSE_ONLINE_IMPORT_MODAL",
  OPEN_LICENSE_MODAL: "OPEN_LICENSE_MODAL",
  CLOSE_LICENSE_MODAL: "CLOSE_LICENSE_MODAL",
  RESET_IMPORT_ONLINE_REDUCER: "RESET_IMPORT_ONLINE_REDUCER",
  SHOW_DIMMED_SCREEN: "SHOW_DIMMED_SCREEN",
  //Project Validation
  RESET_PROJECT_VALIDATION_REDUCER: "RESET_PROJECT_VALIDATION_REDUCER",
  ONLY_SHOW_PROJECT_INFORMATION_SCREEN: "ONLY_SHOW_PROJECT_INFORMATION_SCREEN",
  TOGGLE_PROJECT_VALIDATION_STEPPER: "TOGGLE_PROJECT_VALIDATION_STEPPER",
  GO_TO_PROJECT_VALIDATION_STEP: "GO_TO_PROJECT_VALIDATION_STEP",
  UPDATE_PROJECT_VALIDATION_NEXT_BUTTON_STATUS: "UPDATE_PROJECT_VALIDATION_NEXT_BUTTON_STATUS",
  MISSING_VERSES_CHECK: "MISSING_VERSES_CHECK",
  ADD_PROJECT_VALIDATION_STEP: "ADD_PROJECT_VALIDATION_STEP",
  REMOVE_PROJECT_VALIDATION_STEP: "REMOVE_PROJECT_VALIDATION_STEP",
  // project Details reducer
  UPDATE_SELECTED_PROJECT_FILENAME: "UPDATE_SELECTED_PROJECT_FILENAME",
  UPDATE_SOURCE_PROJECT_PATH: "UPDATE_SOURCE_PROJECT_PATH",
  SAVE_BOOK_ID_AND_BOOK_NAME_IN_MANIFEST: "SAVE_BOOK_ID_AND_BOOK_NAME_IN_MANIFEST",
  SAVE_RESOURCE_ID_IN_MANIFEST: "SAVE_RESOURCE_ID_IN_MANIFEST",
  SAVE_NICKNAME_IN_MANIFEST: "SAVE_NICKNAME_IN_MANIFEST",
  SAVE_LANGUAGE_DETAILS_IN_MANIFEST: "SAVE_LANGUAGE_DETAILS_IN_MANIFEST",
  SAVE_TRANSLATORS_LIST_IN_MANIFEST: "SAVE_TRANSLATORS_LIST_IN_MANIFEST",
  SAVE_CHECKERS_LIST_IN_MANIFEST: "SAVE_CHECKERS_LIST_IN_MANIFEST",
  ADD_MANIFEST_PROPERTY: "ADD_MANIFEST_PROPERTY",
  SET_PROJECT_TYPE: "SET_PROJECT_TYPE",
  OLD_SELECTED_PROJECT_FILENAME: "OLD_SELECTED_PROJECT_FILENAME",
  // Group menu
  TOGGLE_MENU_DRAWER: "TOGGLE_MENU_DRAWER",
  GROUP_MENU_EXPAND_SUBMENU: "GROUP_MENU_EXPAND_SUBMENU",
  GROUP_MENU_SET_FILTER: "GROUP_MENU_TOGGLE_FILTER",
  CLEAR_PREVIOUS_FILTERS: "CLEAR_PREVIOUS_FILTERS",
  // copy right check reducer
  SELECT_PROJECT_LICENSE_ID: "SELECT_PROJECT_LICENSE_ID",
  LOAD_PROJECT_LICENSE_MARKDOWN: "LOAD_PROJECT_LICENSE_MARKDOWN",
  CLEAR_COPYRIGHT_CHECK_REDUCER: "CLEAR_COPYRIGHT_CHECK_REDUCER",
  // project information check reducer
  SET_BOOK_ID_IN_PROJECT_INFORMATION_REDUCER: "SET_BOOK_ID_IN_PROJECT_INFORMATION_REDUCER",
  SET_RESOURCE_ID_IN_PROJECT_INFORMATION_REDUCER: "SET_RESOURCE_ID_IN_PROJECT_INFORMATION_REDUCER",
  SET_NICKNAME_IN_PROJECT_INFORMATION_REDUCER: "SET_NICKNAME_IN_PROJECT_INFORMATION_REDUCER",
  SET_LANGUAGE_ID_IN_PROJECT_INFORMATION_REDUCER: "SET_LANGUAGE_ID_IN_PROJECT_INFORMATION_REDUCER",
  SET_LANGUAGE_NAME_IN_PROJECT_INFORMATION_REDUCER: "SET_LANGUAGE_NAME_IN_PROJECT_INFORMATION_REDUCER",
  SET_LANGUAGE_DIRECTION_IN_PROJECT_INFORMATION_REDUCER: "SET_LANGUAGE_DIRECTION_IN_PROJECT_INFORMATION_REDUCER",
  SET_ALL_LANGUAGE_INFO_IN_PROJECT_INFORMATION_REDUCER: "SET_ALL_LANGUAGE_INFO_IN_PROJECT_INFORMATION_REDUCER",
  SET_CONTRIBUTORS_IN_PROJECT_INFORMATION_REDUCER: "SET_CONTRIBUTORS_IN_PROJECT_INFORMATION_REDUCER",
  SET_CHECKERS_IN_PROJECT_INFORMATION_REDUCER: "SET_CHECKERS_IN_PROJECT_INFORMATION_REDUCER",
  CLEAR_PROJECT_INFORMATION_REDUCER: "CLEAR_PROJECT_INFORMATION_REDUCER",
  // online mode reducer / actions
  RESET_ONLINE_MODE_WARNING_ALERT: "RESET_ONLINE_MODE_WARNING_ALERT",
  // WORD ALIGNMENT DATA
  UPDATE_ALIGNMENT_DATA: "UPDATE_ALIGNMENT_DATA",
  CLEAR_ALIGNMENT_DATA: "CLEAR_ALIGNMENT_DATA",
  //merge conflict
  CLEAR_MERGE_CONFLICTS_REDUCER: "CLEAR_MERGE_CONFLICTS_REDUCER",
  MERGE_CONFLICTS_CHECK: "MERGE_CONFLICTS_CHECK",
  //locale actions
  LOCALE_LOADED: "LOCALE_LOADED",
  // settings actions
  SET_CSV_SAVE_LOCATION: "SET_CSV_SAVE_LOCATION",
  SET_USFM_SAVE_LOCATION: "SET_USFM_SAVE_LOCATION",
  UPDATE_TOOL_SETTINGS: "UPDATE_TOOL_SETTINGS",
  SET_SETTING: "SET_SETTING",
  TOGGLE_SETTING: "TOGGLE_SETTING",
  // invalidated reducer
  SET_INVALIDATED: "SET_INVALIDATED",
  SET_INVALIDATED_CHECKS_TOTAL: "SET_INVALIDATED_CHECKS_TOTAL",
  SET_VERSE_EDITS_TOTAL: "SET_VERSE_EDITS_TOTAL",
  SET_INVALIDATED_ALIGNMENTS_TOTAL: 'SET_INVALIDATED_ALIGNMENTS_TOTAL',
  ADD_TOOL_API: 'ADD_TOOL_API'
};

export default ActionTypes;
