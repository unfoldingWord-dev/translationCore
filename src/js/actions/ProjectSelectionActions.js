import consts from './ActionTypes';
// actions
import * as AlertModalActions from './AlertModalActions';
import * as ToolsMetadataActions from './ToolsMetadataActions';
import * as RecentProjectsActions from './RecentProjectsActions';
import * as BodyUIActions from './BodyUIActions';
import * as ProjectDetailsActions from './projectDetailsActions';
import * as TargetLanguageActions from './TargetLanguageActions';
import * as ProjectValidationActions from './ProjectValidationActions';
// helpers
import * as ProjectSelectionHelpers from '../helpers/ProjectSelectionHelpers';
import * as LoadHelpers from '../helpers/LoadHelpers';
import * as manifestHelpers from '../helpers/manifestHelpers';
import * as usfmHelpers from '../helpers/usfmHelpers';


/**
 * Wrapper function to initate selection of a project from path.
 * @param {string} projectPath - Path location in the filesystem for the project.
 * @param {string} projectLink - Link to the projects git repo if provided i.e. https://git.door43.org/royalsix/fwe_tit_text_reg.git.
 */
export function selectProject(projectPath, projectLink) {
  return ((dispatch, getState) => {
    const { username } = getState().loginReducer.userdata;
    if (!projectPath) {
      return dispatch(AlertModalActions.openAlertDialog("No project path specified"));
    }
    projectPath = LoadHelpers.saveProjectInHomeFolder(projectPath);
    let manifest, params, targetLanguage;
    /**@type {String} */
    let USFMFilePath = usfmHelpers.isUSFMProject(projectPath);
    //If present proceed to usfm loading process
    if (USFMFilePath) {
      let usfmProjectObject = usfmHelpers.getProjectDetailsFromUSFM(USFMFilePath, projectPath);
      let { parsedUSFM, direction } = usfmProjectObject;
      targetLanguage = parsedUSFM;
      manifest = usfmHelpers.getUSFMProjectManifest(projectPath, projectLink, parsedUSFM, direction, username);
      params = usfmHelpers.getUSFMParams(projectPath, manifest);
    } else {
      //If no usfm file found proceed to load regular loading process
      manifest = ProjectSelectionHelpers.getProjectManifest(projectPath, projectLink, username);
      if (!manifest) dispatch(AlertModalActions.openAlertDialog("No valid manifest found in project"));
      params = manifestHelpers.getParams(projectPath, manifest);
    }
    dispatch(clearLastProject());
    dispatch(loadProjectDetails(projectPath, manifest, params));
    dispatch(ProjectValidationActions.validateProject((isValidProject) => {
      if (isValidProject) {
        TargetLanguageActions.generateTargetBible(projectPath, targetLanguage, manifest);
        dispatch(displayTools());
      } else {
        dispatch(ProjectValidationActions.showStepper(true));
      }
    }));
    //TODO: Factor back into project opening workflow
    // if (LoadHelpers.projectHasMergeConflicts(projectPath, manifest.project.id)) dispatch(AlertModalActions.openAlertDialog("Oops! The project you are trying to load has a merge conflict and cannot be opened in this version of translationCore! Please contact Help Desk (help@door43.org) for assistance."));
    // if (LoadHelpers.projectIsMissingVerses(projectPath, manifest.project.id)) {
    //   dispatch(confirmOpenMissingVerseProjectDialog(projectPath, manifest))
    // } else {
    //   dispatch(displayTools(manifest));
    // }
  })
}

/**
 * @description - This action creates a confirm dialog that ensures
 * if the user wants to use a project with missing verses
 * @param {string} projectPath - path location in the filesystem for the project.
 * @param {object} manifest project manifest.
 */
export function confirmOpenMissingVerseProjectDialog(projectPath, manifest) {
  return ((dispatch) => {
    const callback = (option) => {
      dispatch(AlertModalActions.closeAlertDialog());
      if (option != "Cancel") {
        dispatch(displayTools());
      } else {
        dispatch(clearLastProject());
      }
    }
    dispatch(AlertModalActions.openOptionDialog(
      "Oops! Your project has blank verses! Please contact Help Desk (help@door43.org) for assistance with fixing this problem. If you proceed without fixing, some features may not work properly",
      callback,
      "Continue Without Fixing",
      "Cancel"
    ));
  })
}

/**
 * @description loads and set the projects details into the projectDetailsReducer.
 * @param {string} projectPath - path location in the filesystem for the project.
 * @param {object} manifest - project manifest.
 * @param {object} params - parameters defining a projects detals, similiar to metadata.
 */
export function loadProjectDetails(projectPath, manifest, params) {
  return ((dispatch) => {
    LoadHelpers.migrateAppsToDotApps(projectPath);
    dispatch(ProjectDetailsActions.setSaveLocation(projectPath));
    dispatch(ProjectDetailsActions.setProjectManifest(manifest));
    dispatch(ProjectDetailsActions.setProjectDetail("bookName", manifest.project.name));
    dispatch(ProjectDetailsActions.setProjectParams(params));
  });
}

export function clearLastProject() {
  return ((dispatch) => {
    dispatch(BodyUIActions.toggleHomeView(true));
    dispatch(ProjectDetailsActions.resetProjectDetail());
    dispatch({ type: consts.CLEAR_PREVIOUS_GROUPS_DATA });
    dispatch({ type: consts.CLEAR_PREVIOUS_GROUPS_INDEX });
    dispatch({ type: consts.CLEAR_CONTEXT_ID });
    dispatch({ type: consts.CLEAR_CURRENT_TOOL_DATA });
    dispatch({ type: consts.CLEAR_RESOURCES_REDUCER });
    dispatch({
      type: consts.SET_CURRENT_TOOL_TITLE,
      currentToolTitle: ""
    });
  });
}

export function displayTools() {
  return ((dispatch, getState) => {
    const { currentSettings } = getState().settingsReducer;
    const { manifest } = getState().projectDetailsReducer;
    if (manifestHelpers.checkIfValidBetaProject(manifest) || currentSettings.developerMode) {
      dispatch(ToolsMetadataActions.getToolsMetadatas());
      // Go to toolsCards page
      dispatch(BodyUIActions.goToStep(3));
    } else {
      dispatch(AlertModalActions.openAlertDialog('You can only load Titus projects for now.'));
      dispatch(RecentProjectsActions.getProjectsFromFolder());
      dispatch(clearLastProject())
    }
  });
}
