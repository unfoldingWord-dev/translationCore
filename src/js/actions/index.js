/**
 * @author Ian Hoegen
 * @description: This serves as the new store and events api.
 * @usage:
 * Actions.sendAction('YOUR_EVENT', data) // Emits an event of type YOUR_EVENT, adds data to store
 * Actions.getData('YOUR_EVENT') // Retrieves data from the event emitted
 ***********************************************************************************/

const EventEmitter = require('events').EventEmitter;
const Reducers = require('../reducers');

class Actions extends EventEmitter {
  constructor() {
    super();
    this.store = [];
  }
  sendAction(eventName, actionObj) {
    if(Reducers[eventName]) {
      this.store[eventName] = Reducers[eventName](this.store[eventName], actionObj);
    } else {
      this.store[eventName] = actionObj;
    }
    this.emitChange(eventName);
  }

  subscribe(eventName, callback) {
    this.on(eventName, callback);
  }

  unsubscribe(eventName, callback) {
    this.removeListener(eventName, callback);
  }

  emitChange(eventName, data) {
    this.emit(eventName, data);
  }

  getData (actionType) {
    return this.store[actionType];
  }
}

var actions = new Actions();
module.exports = actions;
