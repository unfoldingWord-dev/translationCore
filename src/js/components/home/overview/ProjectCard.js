import React, { Component } from 'react'

class ProjectCard extends Component {
  render() {
    return (
      <div>
        ProjectCard
      </div>
    )
  }
}

export default ProjectCard
