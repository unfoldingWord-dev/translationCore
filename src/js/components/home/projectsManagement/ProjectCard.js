import React from 'react';
import PropTypes from 'prop-types';
import { Glyphicon } from 'react-bootstrap';
// components
import TemplateCard from '../TemplateCard';
import ProjectCardMenu from './ProjectCardMenu';
import Hint from '../../Hint';

let ProjectCard = (props) => {
  const { projectName, projectSaveLocation, accessTimeAgo, bookAbbr, bookName, target_language, isSelected } = props.projectDetails;
  let cardDetails = [
    {
      glyph: 'time',
      text: accessTimeAgo
    },
    {
      glyph: 'book',
      text: bookName && bookAbbr ? bookName + ' (' + bookAbbr + ')' : 'No book information found'
    },
    {
      glyph: 'globe',
      text: target_language.name && target_language.id ? target_language.name + ' (' + target_language.id + ')' : 'No language information found'
    }
  ];

  // content
  let content = (
    <div style={{ display: 'flex', justifyContent: 'space-between', marginTop: '-10px' }}>
      <div style={{ display: 'flex', flexDirection: 'column', justifyContent: 'space-between' }}>
        <Hint position={'bottom'} label={projectName}>
          <strong style={{
            fontSize: 'x-large',
            overflow: 'hidden',
            maxWidth: 400,
            textOverflow: 'ellipsis',
            display: 'block',
            whiteSpace: 'nowrap'
          }}> {projectName} </strong>
        </Hint>
        <div style={{ display: 'flex', justifyContent: 'space-between', width: '410px', marginBottom: '6px' }}>
          <table style={{width: "100%"}}>
            <tbody>
              <tr>
              {
                cardDetails.map((cardDetail) => {
                  let width;
                  switch(cardDetail.glyph){
                    case 'globe':
                      width = "40%";
                      break;
                    case 'time':
                    case 'book':
                    default:
                      width = '30%';
                      break;
                  }
                  return (
                    <td style={{width: width}} key={cardDetail.glyph}>
                      <table style={{width: "100%"}}>
                        <tbody>
                          <tr>
                            <td style={{width: "1px"}}>
                              <Glyphicon glyph={cardDetail.glyph} style={{ marginRight: '5px', top: '2px' }} />
                            </td>
                            <td>
                              {cardDetail.text}
                            </td>
                          </tr>
                        </tbody>
                      </table>
                    </td>
                  );
                })
              }
              </tr>
            </tbody>
          </table>
        </div>
      </div>
      <div style={{ display: 'flex', flexDirection: 'column', justifyContent: 'space-between', textAlign: 'right', marginRight: '-6px' }}>
        <ProjectCardMenu projectSaveLocation={projectSaveLocation} {...props} />
        <div>
          <button className='btn-prime' disabled={isSelected} onClick={() => { props.actions.selectProject(projectSaveLocation) }} style={{ width: '90px', marginBottom: '0' }}>
            Select
          </button>
        </div>
      </div>
    </div>
  );

  // heading
  const heading = (
    <span></span>
  );

  return (
    <TemplateCard
      heading={heading}
      content={content}
    />
  );
};

ProjectCard.propTypes = {
  projectDetails: PropTypes.object.isRequired,
  actions: PropTypes.object.isRequired
};

export default ProjectCard;
