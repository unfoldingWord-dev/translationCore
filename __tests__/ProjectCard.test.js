/* eslint-env jest */

import React from 'react';
import ProjectCard from '../src/js/components/home/projectsManagement/ProjectCard';
import renderer from 'react-test-renderer';
import * as ProjectSelectionActions from '../src/js/actions/ProjectSelectionActions';
import MuiThemeProvider from 'material-ui/styles/MuiThemeProvider';
import TranslateIcon from 'material-ui/svg-icons/action/translate';

// Tests for ProjectCard React Component
describe('Test ProjectCard component',()=>{
  test('Comparing ProjectCard Component render with snapshot taken 11/09/2017 in __snapshots__ should match', () => {
    const projectDetails = {
      projectName: 'en_1co_ulb',
      projectSaveLocation: '/tmp/en_1co_ulb',
      accessTimeAgo: '5 days ago',
      bookAbbr: '1co',
      bookName: '1 Corinthians',
      target_language: {
        id: 'en',
        name: 'English'
      },
      isSelected: false
    };
    const actions = {
      selectProject: (projectPath) => {
        dispatch(ProjectSelectionActions.selectProject(projectPath));
      }
    };
    const renderedValue =  renderer.create(
      <MuiThemeProvider>
        <ProjectCard
          user={"johndoe"}
          key={"en_1co"}
          projectDetails={projectDetails}
          actions={actions} />
      </MuiThemeProvider>
    ).toJSON();
    expect(renderedValue).toMatchSnapshot();
  });

  test('Comparing ProjectCard Component should use characters icon for language graphic', () => {
    // given
    const projectDetails = {
      projectName: 'en_1co_ulb',
      projectSaveLocation: '/tmp/en_1co_ulb',
      accessTimeAgo: '5 days ago',
      bookAbbr: '1co',
      bookName: '1 Corinthians',
      target_language: {
        id: 'en',
        name: 'English'
      },
      isSelected: false
    };
    const actions = {
      selectProject: (projectPath) => {
        dispatch(ProjectSelectionActions.selectProject(projectPath));
      }
    };
    const translateSvg = getTranslationIconSvg();
    const expectedIcons = ['glyphicon glyphicon-time','glyphicon glyphicon-book', translateSvg];

    // when
    const renderedValue =  renderer.create(
      <MuiThemeProvider>
        <ProjectCard
          user={"johndoe"}
          key={"en_1co"}
          projectDetails={projectDetails}
          actions={actions} />
      </MuiThemeProvider>
    ).toJSON();

    // then
    validateDetailIcons(renderedValue, expectedIcons);
  });

  //
  // Helpers
  //

  function getTranslationIconSvg() {
    const renderedTranslateIcon = renderer.create(
      <MuiThemeProvider>
        <TranslateIcon style={{height: "20px", width: "20px", color: "#000000", marginRight: '5px', marginTop: '6px'}}/>
      </MuiThemeProvider>
    ).toJSON();
    const iconPath = searchForChildren(renderedTranslateIcon, 'path');
    const translateSvg = iconPath.length ? iconPath[0].props.d : null;
    return translateSvg;
  }

  function validateDetailIcons(renderedValue, expectedIcons) {
    const cardDetails = searchForChildren(renderedValue, 'td');
    expect(cardDetails.length).toEqual(expectedIcons.length);
    for (let i = 0; i < cardDetails.length; i++) {
      let iconFound = null;
      const detail = cardDetails[i];
      const detailLayout = searchForChildren(detail, 'td');
      const iconDetail = detailLayout[0];
      const iconSpan = searchForChildren(iconDetail, 'span');
      if (iconSpan.length) {
        iconFound = iconSpan[0].props.className;
      } else {
        const iconPath = searchForChildren(iconDetail, 'path');
        if (iconPath.length) {
          iconFound = iconPath[0].props.d;
        }
      }
      expect(iconFound).toEqual(expectedIcons[i]);
    }
  }

  /**
   * @description - get text shown on rendered html (json format)
   * @param rendered - rendered html (json format)
   * @return {Array}
   */
  function getDisplayedText(rendered) {
    const displayedText = [];
    rendered.forEach((item) => {
      let text = "";
      if (typeof item === 'string') {
        if ((item.length) && ((item !== " ") && (item !== "\xA0"))) { // ignore " " whitespace
          text = item;
        }
      } else if(Array.isArray(item)) {
        const array_labels = getDisplayedText(item);
        text = array_labels.join('');
      } else if (item.children) {
        text = getDisplayedTextFromChildren(item);
      }
      displayedText.push(text);
    });
    return displayedText;
  }

  /**
   * @description - get text shown on renderedItem
   * @param renderedItem
   * @return {String}
   */
  function getDisplayedTextFromChildren(renderedItem) {
    const child_texts = getDisplayedText(renderedItem.children);
    let text = "";
    child_texts.forEach((child_label) => {
      if (child_label.length) {
        text += '[' + child_label + ']';
      }
    });
    return text;
  }

  /**
   * @description - find children that match findType
   * @param search
   * @param findType
   * @return {Array}
   */
  function searchForChildren(search, findType, depth=0) {
    let found = [];
    if(search._depth_) {
      depth = search._depth_;
    }
    if (search.children) {
      search.children.forEach((child) => {
        if (child.type === findType) {
          child._depth_ = depth+1; // save depth
          found.push(child);
        } else if(child.children) {
          const found_below = searchForChildren(child, findType, depth+1);
          if(found_below.length) {
            found = found.concat(found_below);
          }
        }
      });
    }
    return found;
  }

});
