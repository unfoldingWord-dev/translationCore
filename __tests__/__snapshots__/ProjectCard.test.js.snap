// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test ProjectCard component Comparing ProjectCard Component render with snapshot taken 11/06/2017 in __snapshots__ should match 1`] = `
<div
  style={
    Object {
      "flex": 1,
    }
  }
>
  <span />
  <div
    style={
      Object {
        "WebkitTapHighlightColor": "rgba(0,0,0,0)",
        "background": "white",
        "backgroundColor": "#ffffff",
        "borderRadius": 2,
        "boxShadow": "0 1px 6px rgba(0, 0, 0, 0.12),
         0 1px 4px rgba(0, 0, 0, 0.12)",
        "boxSizing": "border-box",
        "color": "rgba(0, 0, 0, 0.87)",
        "fontFamily": "Roboto, sans-serif",
        "marginTop": "5px",
        "muiPrepared": true,
        "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
        "zIndex": 1,
      }
    }
  >
    <div
      style={
        Object {
          "paddingBottom": 8,
        }
      }
    >
      <div
        style={
          Object {
            "color": "rgba(0, 0, 0, 0.87)",
            "fontSize": 14,
            "muiPrepared": true,
            "padding": 16,
          }
        }
      >
        <div
          style={
            Object {
              "display": "flex",
              "justifyContent": "space-between",
              "marginTop": "-10px",
            }
          }
        >
          <div
            style={
              Object {
                "display": "flex",
                "flexDirection": "column",
                "justifyContent": "space-between",
              }
            }
          >
            <span
              aria-label="en_1co_ulb"
              className="hint--bottom hint--medium"
            >
              <strong
                style={
                  Object {
                    "display": "block",
                    "fontSize": "x-large",
                    "maxWidth": 400,
                    "overflow": "hidden",
                    "textOverflow": "ellipsis",
                    "whiteSpace": "nowrap",
                  }
                }
              >
                 
                en_1co_ulb
                 
              </strong>
            </span>
            <div
              style={
                Object {
                  "display": "flex",
                  "justifyContent": "space-between",
                  "marginBottom": "6px",
                  "width": "410px",
                }
              }
            >
              <table
                style={
                  Object {
                    "width": "100%",
                  }
                }
              >
                <tbody>
                  <tr>
                    <td
                      style={
                        Object {
                          "width": "30%",
                        }
                      }
                    >
                      <table
                        style={
                          Object {
                            "width": "100%",
                          }
                        }
                      >
                        <tbody>
                          <tr>
                            <td
                              style={
                                Object {
                                  "width": "1px",
                                }
                              }
                            >
                              <span
                                className="glyphicon glyphicon-time"
                                style={
                                  Object {
                                    "marginRight": "5px",
                                    "top": "2px",
                                  }
                                }
                              />
                            </td>
                            <td>
                              5 days ago
                            </td>
                          </tr>
                        </tbody>
                      </table>
                    </td>
                    <td
                      style={
                        Object {
                          "width": "30%",
                        }
                      }
                    >
                      <table
                        style={
                          Object {
                            "width": "100%",
                          }
                        }
                      >
                        <tbody>
                          <tr>
                            <td
                              style={
                                Object {
                                  "width": "1px",
                                }
                              }
                            >
                              <span
                                className="glyphicon glyphicon-book"
                                style={
                                  Object {
                                    "marginRight": "5px",
                                    "top": "2px",
                                  }
                                }
                              />
                            </td>
                            <td>
                              1 Corinthians (1co)
                            </td>
                          </tr>
                        </tbody>
                      </table>
                    </td>
                    <td
                      style={
                        Object {
                          "width": "40%",
                        }
                      }
                    >
                      <table
                        style={
                          Object {
                            "width": "100%",
                          }
                        }
                      >
                        <tbody>
                          <tr>
                            <td
                              style={
                                Object {
                                  "width": "1px",
                                }
                              }
                            >
                              <span
                                className="glyphicon glyphicon-globe"
                                style={
                                  Object {
                                    "marginRight": "5px",
                                    "top": "2px",
                                  }
                                }
                              />
                            </td>
                            <td>
                              English (en)
                            </td>
                          </tr>
                        </tbody>
                      </table>
                    </td>
                  </tr>
                </tbody>
              </table>
            </div>
          </div>
          <div
            style={
              Object {
                "display": "flex",
                "flexDirection": "column",
                "justifyContent": "space-between",
                "marginRight": "-6px",
                "textAlign": "right",
              }
            }
          >
            <div
              style={
                Object {
                  "cursor": "pointer",
                }
              }
            >
              <div
                onTouchTap={[Function]}
              >
                <span
                  className="glyphicon glyphicon-option-vertical"
                  style={
                    Object {
                      "fontSize": "large",
                    }
                  }
                />
              </div>
              <div
                style={
                  Object {
                    "display": "none",
                  }
                }
              />
            </div>
            <div>
              <button
                className="btn-prime"
                disabled={false}
                onClick={[Function]}
                style={
                  Object {
                    "marginBottom": "0",
                    "width": "90px",
                  }
                }
              >
                Select
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
